core/src/main/scala/org/apache/spark/config/ConfigEntry.scala

```diff
@@ -0,0 +1,247 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.config
+
+import java.util.concurrent.TimeUnit
+
+import org.apache.spark.SparkConf
+import org.apache.spark.network.util.{ByteUnit, JavaUtils}
+
+/**
+ * An entry contains all meta information for a configuration.
+ *
+ * @param key the key for the configuration
+ * @param defaultValue the default value for the configuration
+ * @param valueConverter how to convert a string to the value. It should throw an exception if the
+ *                       string does not have the required format.
+ * @param stringConverter how to convert a value to a string that the user can use it as a valid
+ *                        string value. It's usually `toString`. But sometimes, a custom converter
+ *                        is necessary. E.g., if T is List[String], `a, b, c` is better than
+ *                        `List(a, b, c)`.
+ * @param doc the document for the configuration
+ * @param isPublic if this configuration is public to the user. If it's `false`, this
+ *                 configuration is only used internally and we should not expose it to the user.
+ * @tparam T the value type
+ */
+private[spark] class ConfigEntry[T] (
+    val key: String,
+    val defaultValue: Option[T],
+    val valueConverter: String => T,
+    val stringConverter: T => String,
+    val doc: String,
+    val isPublic: Boolean) {
+
+  def defaultValueString: String = defaultValue.map(stringConverter).getOrElse("<undefined>")
+
+  /**
+   * Returns a new ConfigEntry that wraps the value in an Option, for config entries that do
+   * not require a value.
+   */
+  def optional: OptionalConfigEntry[T] = {
+    require(!defaultValue.isDefined, s"$this has a default value, cannot be optional.")
+    new OptionalConfigEntry(key, valueConverter, stringConverter, doc, isPublic)
```

Comment:
I agree it may not super intuitive, but it's not only for optional configs. `get(FOO)` may return `String` while `get(BAR)` may return a `Long` and `get(OTHER)` might return `Option[Boolean]`. But read on.

But I disagree that having a `getOption` for `ConfEntry` objects is better. That completely breaks the goal of this change, which is to tie a type to the configuration entry. A configuration entry that has a default value will always return `Some(foo)` for that method, so it doesn't really make a lot of sense to call that method to retrieve that config's value.

An optional config has type `Option[Something]`, not `Something`, which makes way more sense to me. With the current code, if you try to use an optional config as if it were an actual config, you'll get a compile-time error instead of a runtime error telling you the config option is not set.

The last point is why I think the slight awkwardness in having `get` return different types is actually better in the end, because now you have the compiler doing proper type checks instead of relying on your code knowing what the type of a config is.

If you're worried about having to call `.optional` to create optional configs, that can be changed in other ways. For example you could have overloaded methods, one that doesn't take a default value and returns an optional config, and one that does take a default value and returns a different config. Or you could use a builder-style pattern (e.g., `ConfigEntry.intConf(...).withDefault(...)`), although now you have the verbosity the other way.